src/fal/planner/schedule.py

```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Iterator
+
+import networkx as nx
+
+from fal.node_graph import DbtModelNode, NodeGraph, NodeKind
+from fal.planner.tasks import SUCCESS, DBTTask, FalHookTask, FalModelTask, Node
+
+
+def create_node(
+    node: str | nx.DiGraph, properties: dict, node_graph: NodeGraph
+) -> Node:
+    kind = properties["kind"]
+    if isinstance(node, nx.DiGraph):
+        model_ids = sorted(
+            list(node),
+            key=lambda node: node == properties["exit_node"],
+        )
+    else:
+        model_ids = [node]
+
+    if kind is NodeKind.DBT_MODEL:
+        task = DBTTask(model_ids=model_ids)
+    else:
+        assert kind is NodeKind.FAL_MODEL
+        task = FalModelTask(model_ids=model_ids)
+
+    model_node = node_graph.get_node(model_ids[-1])
+    assert model_node is not None
+
+    if isinstance(task, FalModelTask):
+        assert isinstance(model_node, DbtModelNode)
+        task.bound_model = model_node.model
+
+    post_hooks = [
+        FalHookTask(
+            hook_path=hook_path,
+            bound_model=model_node.model,
+        )
+        for hook_path in properties.get("post-hook", [])
+    ]
+    return Node(task, post_hooks=post_hooks)
+
+
+@dataclass
+class NodeQueue:
+    nodes: list[Node]
+    _staged_nodes: list[Node] = field(default_factory=list)
+    _counter: int = 0
+
+    def __bool__(self) -> bool:
+        return bool(self.nodes)
+
+    def _calculate_node_score(self, target_node: Node) -> tuple[int, int]:
+        # Determine the priority of the node by doing a bunch
+        # of calculations. This doesn't really need to be 100% precise,
+        # since if we don't have this we'll have to schedule randomly.
+
+        # 1. Number of nodes which are only waiting this node (direct dependants)
+        # 2. Number of nodes which are waiting this node and other nodes (indirect dependants)
+        # ...
+
+        direct_dependants = 0
+        indirect_dependants = 0
+
+        for node in self.nodes:
+            if node is target_node:
+                continue
+
+            if any(dependency is target_node for dependency in node.dependencies):
+                indirect_dependants += 1
+                if len(node.dependencies) == 1:
+                    direct_dependants += 1
+
+        return (direct_dependants, indirect_dependants)
+
+    def _stage_node(self, target_node: Node) -> None:
+        # When we start running a node, we don't want to simply remove it
+        # just yet (since that would unblock all of its dependencies). So
+        # we temporarily add it to a separate group of nodes (staged).
+
+        self._counter += 1
+        target_node.set_run_index(self._counter)
+
+        self.nodes.remove(target_node)
+        self._staged_nodes.append(target_node)
```

Comment:
Is it possible that a race condition could tink a dependency is unblocked? Should we swap the order of these 2?

Reply:
Same as question as below.

Reply:
I see now that `_staged_nodes` is not really used to check if a node is in process or not. `_staged_nodes` really does not do much, we could not have it at all. Unless it is useful for debugging, maybe.

Reply:
Yep, that is the primary use case (a bit generalized, like not making a node disappear all together until it is finished).